feat(waitForElement): a wrapper around MutationObserver

sompylasar opened the pull request:
* Add `waitForElement` implementation, tests, docs
* Add `mutationobserver-shim` to `dependencies` to provide
`MutationObserver` in jest tests where `jsdom` has no built-in support
for it: https://github.com/jsdom/jsdom/issues/639
* Update @sompylasar contributions

kentcdodds commented on README.md:
````diff
@@ -305,6 +306,50 @@ The default `interval` is `50ms`. However it will run your callback immediately
+Defined as:
+
+```typescript
+function waitForElements<T>(
````
Normally I think people will be looking for a single element. Should this be called `waitForElement`?

In addition, we should show an example (in the FAQ?) of getting an array of elements:

```javascript
const elements = waitForElement(() => [
  getByText('username'),
  getByText('password')
])
```

kentcdodds commented on README.md:
````diff
@@ -305,6 +306,50 @@ The default `interval` is `50ms`. However it will run your callback immediately
+// ...
+```
+
+Using `MutationObserver` is more efficient than polling the DOM at regular intervals with `wait`.
````
Let's add a link to the MDN docs for this so folks know that it's actually a browser feature. Most people have probably never heard of it. I only heard of it a few weeks ago when building the cypress command waiter.

kentcdodds commented on README.md:
```diff
@@ -305,6 +306,50 @@ The default `interval` is `50ms`. However it will run your callback immediately
+Using `MutationObserver` is more efficient than polling the DOM at regular intervals with `wait`.
+
+The default `callback` is a no-op function (used like `await waitForElements()`). This can
+be helpful if you only need to wait for the next DOM change (see `mutationObserverOptions` to learn which changes are detected).
```
Add a link to `mutationObserverOptions` :)

kentcdodds commented on README.md:
```diff
@@ -305,6 +306,50 @@ The default `interval` is `50ms`. However it will run your callback immediately
+
+The default `mutationObserverOptions` is `{subtree: true, childList: true}` which will detect
+additions and removals of child elements (including text nodes) in the `container` and any of its descendants.
+It won't detect attribute changes unless you add `attributes: true` to the options.
```
Perfect :+1:

kentcdodds commented on package.json:
```diff
@@ -39,7 +39,8 @@
     "jest-in-case": "^1.0.2",
-    "kcd-scripts": "^0.36.1"
+    "kcd-scripts": "^0.36.1",
+    "mutationobserver-shim": "^0.3.2"
```
This should be a regular dependency.

kentcdodds commented on src/waitForElements.js:
```diff
@@ -0,0 +1,43 @@
+) {
+  return new Promise((resolve, reject) => {
+    // Disabling eslint prefer-const below: either prefer-const or no-use-before-define triggers.
+    let lastError, observer, timer // eslint-disable-line prefer-const
```
Really odd that both of these rules trigger 🤔 I don't see how this code is in violation. An ESLint bug?

sompylasar replied:
If I have it as `const`, I have to initialize it at definition, i.e. at this line:
```
timer = setTimeout(onTimeout, timeout)
```
but then:
```
    function onDone(error, result) {
      clearTimeout(timer)
```
is a usage before definition of `timer`

kentcdodds commented on src/waitForElements.js:
```diff
@@ -0,0 +1,43 @@
+    timer = setTimeout(onTimeout, timeout)
+    observer = new MutationObserver(onMutation)
+    observer.observe(container, mutationObserverOptions)
+  })
```
Could we run `onMutation` above this line? That way if there's no reason to wait for a mutation to proceed we don't have to wait for one.

sompylasar commented:
> So I guess we're not providing them with the mutation observer polyfill ourselves? They have to install that? I'd rather polyfill it for people I think.

We could bundle it I guess, yes.

sompylasar commented:
@kentcdodds All requested changes addressed here: https://github.com/kentcdodds/dom-testing-library/pull/10/commits/1f46848a95735fd8c13a167b72f5b7109decf02d

kentcdodds commented on README.md:
```diff
@@ -305,6 +306,62 @@ The default `interval` is `50ms`. However it will run your callback immediately
+The default `callback` is a no-op function (used like `await waitForElement()`). This can
+be helpful if you only need to wait for the next DOM change (see [`mutationObserverOptions`](#mutationobserveroptions) to learn which changes are detected).
+
+The default `container` is the global `document`. Make sure the elements you wait for will be attached to it, or set a different `container`.
```
Could you add a simple example of this. It's probably the most common use case

kentcdodds replied:
Or maybe just make the example above use it like that.

sompylasar commented:
Turned out the no-argument `waitForElement()` stopped working as advertised after we've added the synchronous `onMutation()` call. Fixed it here: https://github.com/kentcdodds/dom-testing-library/pull/10/commits/88bab8733cbbf393404fc3fd37062163348cd2c1

Added `container` option to the examples here: https://github.com/kentcdodds/dom-testing-library/pull/10/commits/b7158cf1b497043dff526ad0b31de31d566a508c

kentcdodds commented on src/index.js:
```diff
@@ -6,4 +6,5 @@ export {queries}
 
 export * from './queries'
 export * from './wait'
+export * from './waitForElement'
```
This is a bit of a nit-pick and I really should have an eslint rule for this, but due to issues with file casing + different operating systems + git, I prefer filenames to always use kebab-case rather than camelCase. Could you update the files to use kebab-case please?

sompylasar replied:
Okay, but tell this to React team. It's for the sake of searchability that the file is literally named the same as the thing it contains and exports. Better have a lint rule that ensures the names strictly match case with the filesystem (and there is such, I think in the eslint-plugin-import).

kentcdodds replied:
I have told them and they rejected it. But I've had too many people bitten by file casing issues across operating systems to do anything else.

kentcdodds replied:
If searchability is key then you say: "Oh, here's a thing called `waitForElement` in the code, so I'll turn that to kebab-case and search for `wait-for-element`." Believe me, I've felt these pains before as well. We just have different solutions.

In addition, it's why I'm starting to avoid `default export`s now because the greppability of a named export is better.

sompylasar replied:
I just select the thing, cmd+shift+f, no need to convert. Especially when doing large refactorings. Default exports I agree can become messy if coders don't follow to name them the same at import site, it's a matter of discipline which is not enforced by the program structure.

sompylasar replied:
👉  https://github.com/kentcdodds/dom-testing-library/pull/10/commits/1252d0824d454b27ff8ed92390af681984069668

kentcdodds commented on src/__tests__/waitForElement.js:
```diff
@@ -0,0 +1,278 @@
+  expect(container).toMatchSnapshot()
+  expect(testEl.parentNode).toBe(container)
+
+  await promise
```
Rather than `await promise`, I'd prefer `return promise`. Maybe I'm being irrational here, but it feels better to me.

sompylasar replied:
I'm changing this, but if this line was the only one with `await` expression, then `eslint` would say `async` function must have an `await`. We are lucky that there are more `await`s in all the tests.

P.S. I don't agree with this rule because `async` functions have their use for the error handling; `async` function is just the one that is wrapped in a `return new Promise(...)`.

sompylasar replied:
👉 https://github.com/kentcdodds/dom-testing-library/pull/10/commits/414a3dca4f75abb64033d7e614255b66e42bf694

kentcdodds commented on src/__tests__/waitForElement.js:
```diff
@@ -0,0 +1,278 @@
+}
+
+async function skipSomeTimeForMutationObserver(delayMs = 50) {
+  // Using `setTimeout` >30ms instead of `wait` here because `mutationobserver-shim` uses `setTimeout` ~30ms.
```
Thanks for this comment :+1:

sompylasar replied:
You're welcome 😉

kentcdodds commented on src/__tests__/waitForElement.js:
```diff
@@ -0,0 +1,278 @@
+  expect(errorHandler).toHaveBeenCalledTimes(0)
+
+  container.setAttribute('data-test-attribute', 'something changed once')
+  await skipSomeTimeForMutationObserver(200)
```
Does this time need to be so long? I'd prefer it to be as short as possible so these tests run quickly :)

sompylasar replied:
No it doesn't, will reduce to 50.

sompylasar replied:
👉 https://github.com/kentcdodds/dom-testing-library/pull/10/commits/edfb24144f1bc1a34dfa4d3e6e91ff1b45a5d81d

kentcdodds commented on src/__tests__/waitForElement.js:
```diff
@@ -0,0 +1,278 @@
+  expect(errorHandler).toHaveBeenCalledTimes(0)
+
+  container.setAttribute('data-test-attribute', 'something changed twice')
+  await skipSomeTimeForMutationObserver(150)
```
Same here.

kentcdodds commented on src/__tests__/waitForElement.js:
```diff
@@ -0,0 +1,278 @@
+  expect(errorHandler).toHaveBeenCalledTimes(0)
+
+  container.setAttribute('data-test-attribute', 'something changed once')
+  await skipSomeTimeForMutationObserver(200)
```
And here

kentcdodds commented on src/__tests__/waitForElement.js:
```diff
@@ -0,0 +1,278 @@
+  expect(errorHandler).toHaveBeenCalledTimes(0)
+
+  container.setAttribute('data-test-attribute', 'something changed twice')
+  await skipSomeTimeForMutationObserver(150)
```
And here :)

kentcdodds commented on src/__tests__/waitForElement.js:
```diff
@@ -0,0 +1,278 @@
+  expect(errorHandler).toHaveBeenCalledTimes(0)
+
+  container.setAttribute('data-test-attribute', 'something changed once')
+  await skipSomeTimeForMutationObserver(200)
```
Annnnd here

sompylasar commented:
@kentcdodds Thanks for merging! I'm currently more interested in having `cypress-testing-library` updated to use `waitForElement` and published. I'm already using the updated Cypress code with `waitForElement` in my project (and it works!), but it would be nice to install this piece of code via `cypress-testing-library`.

gnapse commented:
Wow `waitForElement` looks awesome! I did not know something like `MutationObserver` existed.

kentcdodds commented:
> I'm currently more interested in having cypress-testing-library updated to use waitForElement and published.

Sounds great! I'll be happy to set that project up with auto-releases and such as soon as it's thoroughly tested :+1: I've got the tooling all set up so that should be pretty seamless :+1:

sompylasar commented:
Can't promise that, now I only have time to make the product app I'm working on thoroughly tested with Cypress.

kentcdodds commented:
Well, if you can't get cypress-testing-library well tested, then feel free to just get it in a working state and I'd happily manually publish a `0.0.0` so you can at least install and use it until we can get a working version :+1:

sompylasar commented:
@kentcdodds Deal!
